src/stan/model/model_base_crtp.hpp

```diff
@@ -0,0 +1,216 @@
+#ifndef STAN_MODEL_MODEL_BASE_CRTP_HPP
+#define STAN_MODEL_MODEL_BASE_CRTP_HPP
+
+#include <stan/model/model_base.hpp>
+#include <iostream>
+#include <utility>
+#include <vector>
+
+namespace stan {
+namespace model {
+
+/**
+ * Base class employing the curiously recursive template pattern for
+ * static inheritance to adapt templated `log_prob` and `write_array`
+ * methods to their untemplated virtual counterparts declared in
+ * `model_base`.
+ *
+ * The derived class `M` is required to implement the following two
+ * pairs of template functions,
+ *
+ * ```
+ * template <bool propto, bool jacobian, typename T>
+ * T log_prob(std::vector<T>& params_r,
+ *              std::vector<int>& params_i,
+ *              std::ostream* msgs = 0) const;
+
+ * template <bool propto, bool jacobian, typename T>
+ * T log_prob(Eigen::Matrix<T, -1, 1>& params_r,
+ *            std::ostream* msgs = 0) const;
+ * ```
+ *
+ * and
+ *
+ * ```
+ *  template <typename RNG>
+ *    void write_array(RNG& base_rng,
```

Comment:
Also alignment here... delete two spaces in L36-42.

![image](https://user-images.githubusercontent.com/425751/61265508-71d41680-a75e-11e9-834d-f0ce81586781.png)

Reply:
done